# Phase 2: Data Preprocessing & Leakage-Free Feature Pipeline

**Project**: Machine Learning Based Campaign Response Prediction for Bank Marketing  
**Phase**: 02 – Data Engineering & Preprocessing  
**Owner**: Athukorala M. B. (IT24101937)  
**Governance & Standards**: Aligned with Constitution Principles I–VI and [ADR-001 (Validation Split Strategy)](../docs/decision_log.md)

### Purpose
This notebook implements the reproducible, leakage-free data engineering and feature preprocessing pipeline for the UCI Bank Marketing dataset (`bank-additional-full.csv`). It resolves the validation split strategy discrepancy between the initial project proposal and implemented pipeline via an empirical investigation, establishes strict fit-on-train data isolation, applies modular Scikit-Learn transformers, and exports serialized artifacts for downstream model training.

### Inputs
- Raw Dataset: `data/raw/bank-additional-full.csv` (41,188 records, 20 features + target `y`)
- Modular Transformer: `src.transformers.BankFeatureEngineer`

### Outputs
- `data/processed/X_train_processed.csv`: Transformed training feature matrix (32,950 × 64)
- `data/processed/X_test_processed.csv`: Transformed holdout test feature matrix (8,238 × 64)
- `data/processed/y_train.csv`: Training target vector (32,950 records)
- `data/processed/y_test.csv`: Holdout target vector (8,238 records)
- `data/processed/preprocessor_pipeline.joblib`: Serialized fitted `Pipeline`
- `data/processed/feature_metadata.json`: Feature catalog and pipeline specifications

In [1]:
import os
import sys
import json
from pathlib import Path

# Deterministically resolve project root independent of caller working directory
current_dir = Path.cwd().resolve()
if (current_dir / "src" / "transformers.py").exists():
    PROJECT_ROOT = current_dir
elif (current_dir.parent / "src" / "transformers.py").exists():
    PROJECT_ROOT = current_dir.parent
else:
    PROJECT_ROOT = next(
        (p for p in [current_dir] + list(current_dir.parents) if (p / "src" / "transformers.py").exists()),
        current_dir
    )

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import joblib

from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import (
    OneHotEncoder,
    OrdinalEncoder,
    StandardScaler,
    RobustScaler
)

# Plotting & display configuration
sns.set_theme(style="whitegrid", palette="muted")
plt.rcParams.update({'figure.autolayout': True, 'font.sans-serif': 'Arial'})
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

print(f"Project root resolved to: {PROJECT_ROOT}")
print(f"NumPy version:        {np.__version__}")
print(f"Pandas version:       {pd.__version__}")
import sklearn
print(f"Scikit-Learn version: {sklearn.__version__}")
print(f"Joblib version:       {joblib.__version__}")
print("Environment initialized successfully.")

Project root resolved to: /Users/moni/Desktop/ML proj
NumPy version:        2.5.3
Pandas version:       3.0.6
Scikit-Learn version: 1.9.1
Joblib version:       1.6.0
Environment initialized successfully.


/Users/moni/.matplotlib is not a writable directory
Matplotlib created a temporary cache directory at /var/folders/21/5_d9lvkx55j2xppyx13xbhjh0000gn/T/matplotlib-glsm5bp6 because there was an issue with the default path (/Users/moni/.matplotlib); it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
Matplotlib is building the font cache; this may take a moment.


## 1. Raw Data Ingestion & Target Variable Isolation

In [2]:
# Locate dataset dynamically using PROJECT_ROOT anchor
candidates = [
    PROJECT_ROOT / "data" / "raw" / "bank-additional-full.csv",
    Path("../data/raw/bank-additional-full.csv"),
    Path("data/raw/bank-additional-full.csv"),
    Path("bank-additional-full.csv"),
    Path("../../data/raw/bank-additional-full.csv")
]

raw_data_path = None
for p in candidates:
    if p.exists():
        raw_data_path = p.resolve()
        break

if raw_data_path is None:
    raise FileNotFoundError("Could not locate 'bank-additional-full.csv' in expected directories.")

print(f"Loading raw dataset from: {raw_data_path}")
df_raw = pd.read_csv(raw_data_path, sep=';')
print(f"Dataset Dimensions: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns")
print("\nFirst 3 records:")
print(df_raw.head(3))

Loading raw dataset from: /Users/moni/Desktop/ML proj/data/raw/bank-additional-full.csv
Dataset Dimensions: 41,188 rows × 21 columns

First 3 records:
   age        job  marital    education  default housing loan    contact month day_of_week  duration  campaign  pdays  previous     poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y
0   56  housemaid  married     basic.4y       no      no   no  telephone   may         mon       261         1    999         0  nonexistent           1.1          93.994          -36.4      4.857       5191.0  no
1   57   services  married  high.school  unknown      no   no  telephone   may         mon       149         1    999         0  nonexistent           1.1          93.994          -36.4      4.857       5191.0  no
2   37   services  married  high.school       no     yes   no  telephone   may         mon       226         1    999         0  nonexistent           1.1          93.994          -36.4      4.857       5191

In [3]:
# 1. Target encoding: 'yes' -> 1, 'no' -> 0
target_col = 'y'
y = df_raw[target_col].map({'no': 0, 'yes': 1}).astype(int)

# 2. Drop target and leakage variable 'duration'
leakage_col = 'duration'
X = df_raw.drop(columns=[target_col, leakage_col])

print(f"Target variable '{target_col}' encoded as binary (0/1):")
print(y.value_counts(normalize=True).rename('proportion').to_frame().assign(count=y.value_counts()))

print(f"\nFeature Matrix X shape after discarding '{leakage_col}': {X.shape}")
print(f"Is '{leakage_col}' present in X? {'duration' in X.columns}")

Target variable 'y' encoded as binary (0/1):
   proportion  count
y                   
0    0.887346  36548
1    0.112654   4640

Feature Matrix X shape after discarding 'duration': (41188, 19)
Is 'duration' present in X? False


## 2. Validation Split Strategy Investigation & Rationale

### Scope Discrepancy Context
The initial project submission proposed a validation methodology that considers the chronological ordering of the campaign data to avoid leaking time-related patterns into holdout sets. However, before establishing our production preprocessing pipeline, we conducted an empirical investigation comparing:
- **Option A – Temporal Holdout**: Partitioning the dataset chronologically (first 80% development, last 20% holdout).
- **Option B – Stratified Random Holdout**: Partitioning via stratified random sampling (80% development, 20% holdout with `random_state=42`).

The comparison below details why Option B is retained and formally ratified in `docs/decision_log.md` (ADR-001).

In [4]:
# Empirical Investigation: Option A (Temporal Split) vs Option B (Stratified Random Split)

# Option A: Chronological 80/20 split
split_idx = int(len(df_raw) * 0.8)
train_chrono = df_raw.iloc[:split_idx]
test_chrono = df_raw.iloc[split_idx:]

y_chrono_train = (train_chrono['y'] == 'yes').astype(int)
y_chrono_test = (test_chrono['y'] == 'yes').astype(int)

# Option B: Stratified 80/20 split (for comparison)
X_dummy_tr, X_dummy_te, y_strat_tr, y_strat_te = train_test_split(
    X, y, test_size=0.20, random_state=RANDOM_STATE, stratify=y
)

print("=" * 80)
print("EMPIRICAL COMPARISON: TEMPORAL VS STRATIFIED RANDOM SPLIT (80/20)")
print("=" * 80)

# 1. Target class distribution comparison
print("\n1. TARGET CLASS BASE RATE (PRIOR):")
print(f" - Full Dataset Positives:     {y.mean():.4%} ({y.sum():,} / {len(y):,})")
print(f" - Option A (Temporal) Train:  {y_chrono_train.mean():.4%} ({y_chrono_train.sum():,} / {len(y_chrono_train):,})")
print(f" - Option A (Temporal) Test:   {y_chrono_test.mean():.4%} ({y_chrono_test.sum():,} / {len(y_chrono_test):,})")
print(f"   --> Relative Prior Shift:   +{(y_chrono_test.mean() / y_chrono_train.mean() - 1):.1%}")
print(f" - Option B (Stratified) Train:{y_strat_tr.mean():.4%} ({y_strat_tr.sum():,} / {len(y_strat_tr):,})")
print(f" - Option B (Stratified) Test: {y_strat_te.mean():.4%} ({y_strat_te.sum():,} / {len(y_strat_te):,})")

# 2. Covariate Shift Comparison (Macroeconomic Indicators)
econ_vars = ['euribor3m', 'nr.employed', 'emp.var.rate']
print("\n2. MACROECONOMIC COVARIATE SHIFT (OPTION A TEMPORAL):")
for col in econ_vars:
    tr_min, tr_max, tr_mean = train_chrono[col].min(), train_chrono[col].max(), train_chrono[col].mean()
    te_min, te_max, te_mean = test_chrono[col].min(), test_chrono[col].max(), test_chrono[col].mean()
    print(f" - {col:14s} | Train: [{tr_min:.3f}, {tr_max:.3f}] (mean {tr_mean:.3f}) | Test: [{te_min:.3f}, {te_max:.3f}] (mean {te_mean:.3f})")

# 3. Categorical Support Coverage
train_months = set(train_chrono['month'].unique())
test_months = set(test_chrono['month'].unique())
missing_in_train = test_months - train_months
print(f"\n3. CATEGORICAL SUPPORT:")
print(f" - Months in Temporal Train: {sorted(list(train_months))}")
print(f" - Months in Temporal Test:  {sorted(list(test_months))}")
if missing_in_train:
    print(f" - WARNING: Month category missing from training set under Option A: {missing_in_train}")

print("=" * 80)

EMPIRICAL COMPARISON: TEMPORAL VS STRATIFIED RANDOM SPLIT (80/20)

1. TARGET CLASS BASE RATE (PRIOR):
 - Full Dataset Positives:     11.2654% (4,640 / 41,188)
 - Option A (Temporal) Train:  6.3733% (2,100 / 32,950)
 - Option A (Temporal) Test:   30.8327% (2,540 / 8,238)
   --> Relative Prior Shift:   +383.8%
 - Option B (Stratified) Train:11.2656% (3,712 / 32,950)
 - Option B (Stratified) Test: 11.2649% (928 / 8,238)

2. MACROECONOMIC COVARIATE SHIFT (OPTION A TEMPORAL):
 - euribor3m      | Train: [1.299, 5.045] (mean 4.269) | Test: [0.634, 1.299] (mean 1.031)
 - nr.employed    | Train: [5099.100, 5228.100] (mean 5195.140) | Test: [4963.600, 5099.100] (mean 5054.626)
 - emp.var.rate   | Train: [-1.800, 1.400] (mean 0.650) | Test: [-3.400, -1.100] (mean -2.192)

3. CATEGORICAL SUPPORT:
 - Months in Temporal Train: ['apr', 'aug', 'dec', 'jul', 'jun', 'mar', 'may', 'nov', 'oct']
 - Months in Temporal Test:  ['apr', 'aug', 'dec', 'jul', 'jun', 'mar', 'may', 'nov', 'oct', 'sep']
 - WARNING:

### Summary of Findings & Decision (ADR-001)

The empirical analysis reveals three critical risks associated with Option A (Temporal Holdout):
1. **Severe Covariate Shift**: The global financial crisis (2008–2010) caused `euribor3m`, `nr.employed`, and `emp.var.rate` to experience non-overlapping domains between the development and holdout periods. Machine learning models (e.g. tree ensembles) cannot extrapolate into unobserved covariate ranges.
2. **Extreme Concept Drift / Prior Shift**: The target conversion rate jumps by **+384%** (from $6.37\%$ to $30.83\%$), with over $54.7\%$ of all positive responders in the entire dataset clustered in the holdout partition. This severely skews probability calibration and contact ranking thresholds.
3. **Categorical Support Discontinuity**: Campaign month `month='sep'` ($570$ records) has 0 training instances in Option A.

**Ratified Strategy**: In accordance with **ADR-001** (recorded in `docs/decision_log.md`), the project retains **Option B – Stratified Random Holdout** (`test_size=0.20, random_state=42, stratify=y`). This guarantees uniform macroeconomic representation, balanced class priors, and valid cross-validation for our primary business lens of customer response ranking.

In [5]:
# Ratified Execution: Stratified 80/20 train/test split with contract verification
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=RANDOM_STATE,
    stratify=y
)

# Contract Invariant Assertions (split-validation-contract.md)
assert len(X_train) == 32950, f"Expected 32,950 train records, got {len(X_train)}"
assert len(X_test) == 8238, f"Expected 8,238 test records, got {len(X_test)}"
assert len(y_train) == 32950
assert len(y_test) == 8238
assert len(X_train) + len(X_test) == 41188

# Mutual Exclusivity Assertion
train_idx_set = set(X_train.index)
test_idx_set = set(X_test.index)
assert len(train_idx_set.intersection(test_idx_set)) == 0, "Index overlap detected between train and test sets!"

# Stratification Balance Assertion (< 0.1% delta)
strat_delta = abs(y_train.mean() - y_test.mean())
assert strat_delta < 0.001, f"Stratification mismatch: delta={strat_delta:.6f}"

print("Data Split Summary (Ratified Option B):")
print(f" - Development Set (X_train): {X_train.shape[0]:,} records ({X_train.shape[0]/len(X):.1%})")
print(f" - Holdout Test Set (X_test):   {X_test.shape[0]:,} records ({X_test.shape[0]/len(X):.1%})")
print(f"\nTarget Class Balance Verification:")
print(f" - Full Dataset Positives:      {y.mean():.4%} ({y.sum():,} / {len(y):,})")
print(f" - Train Set Positives:         {y_train.mean():.4%} ({y_train.sum():,} / {len(y_train):,})")
print(f" - Test Set Positives:          {y_test.mean():.4%} ({y_test.sum():,} / {len(y_test):,})")
print(f" - Stratification Delta:        {strat_delta:.6f}")
print("Stratification successfully verified and contract invariants passed.")

Data Split Summary (Ratified Option B):
 - Development Set (X_train): 32,950 records (80.0%)
 - Holdout Test Set (X_test):   8,238 records (20.0%)

Target Class Balance Verification:
 - Full Dataset Positives:      11.2654% (4,640 / 41,188)
 - Train Set Positives:         11.2656% (3,712 / 32,950)
 - Test Set Positives:          11.2649% (928 / 8,238)
 - Stratification Delta:        0.000007
Stratification successfully verified and contract invariants passed.


## 3. Modular Feature Engineering (`src.transformers`)

In [6]:
# Ensure project root is in sys.path for modular imports
import sys
from pathlib import Path

# Deterministically ensure PROJECT_ROOT is defined before use
if 'PROJECT_ROOT' not in globals():
    current_dir = Path.cwd().resolve()
    PROJECT_ROOT = next(
        (p for p in [current_dir] + list(current_dir.parents) if (p / 'src' / 'transformers.py').exists()),
        current_dir
    )

for p in ['.', '..', str(PROJECT_ROOT)]:
    if (Path(p) / 'src').exists() and str(p) not in sys.path:
        sys.path.insert(0, str(p))

from src.transformers import BankFeatureEngineer

# Test feature engineer transformer on X_train only
fe_test = BankFeatureEngineer()
X_train_fe_sample = fe_test.transform(X_train)

print("Engineered feature verification:")
print(" - New columns added:", [c for c in X_train_fe_sample.columns if c not in X_train.columns])
print(" - Columns removed:   ", [c for c in X_train.columns if c not in X_train_fe_sample.columns])
print("\n'pdays_group' breakdown:")
print(X_train_fe_sample['pdays_group'].value_counts())
print("\n'age_group' breakdown:")
print(X_train_fe_sample['age_group'].value_counts())

Engineered feature verification:
 - New columns added: ['previously_contacted', 'pdays_group', 'age_group']
 - Columns removed:    ['pdays']

'pdays_group' breakdown:
pdays_group
not_contacted    31724
0_to_6_days        915
7_to_14_days       256
15_plus_days        55
Name: count, dtype: int64

'age_group' breakdown:
age_group
30-39    13486
40-49     8430
50-59     5511
<30       4574
60+        949
Name: count, dtype: int64


## 4. End-to-End Scikit-Learn Preprocessing Pipeline Architecture

In [7]:
# 1. Feature column definitions
nominal_cols = [
    'job', 'marital', 'default', 'housing', 'loan',
    'contact', 'month', 'day_of_week', 'poutcome',
    'age_group', 'pdays_group'
]

ordinal_cols = ['education']
education_hierarchy = [[
    'unknown', 'illiterate', 'basic.4y', 'basic.6y', 'basic.9y',
    'high.school', 'professional.course', 'university.degree'
]]

robust_num_cols = ['campaign', 'previous']
standard_num_cols = [
    'age', 'emp.var.rate', 'cons.price.idx',
    'cons.conf.idx', 'euribor3m', 'nr.employed'
]
passthrough_cols = ['previously_contacted']

# 2. Build ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('nominal', OneHotEncoder(handle_unknown='ignore', sparse_output=False), nominal_cols),
        ('ordinal', OrdinalEncoder(categories=education_hierarchy, handle_unknown='use_encoded_value', unknown_value=-1), ordinal_cols),
        ('robust_num', RobustScaler(), robust_num_cols),
        ('standard_num', StandardScaler(), standard_num_cols),
        ('passthrough', 'passthrough', passthrough_cols)
    ],
    remainder='drop',
    verbose_feature_names_out=False
)

# 3. Assemble full preprocessing pipeline
full_pipeline = Pipeline([
    ('feature_engineer', BankFeatureEngineer()),
    ('preprocessor', preprocessor)
])

print("Full Pipeline Architecture:")
print(full_pipeline)

Full Pipeline Architecture:
Pipeline(steps=[('feature_engineer', BankFeatureEngineer()),
                ('preprocessor',
                 ColumnTransformer(transformers=[('nominal',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False),
                                                  ['job', 'marital', 'default',
                                                   'housing', 'loan', 'contact',
                                                   'month', 'day_of_week',
                                                   'poutcome', 'age_group',
                                                   'pdays_group']),
                                                 ('ordinal',
                                                  OrdinalEncoder(categories=[['unknown',
                                                                              'illitera...
                     

## 5. Strict Fit-on-Train Pipeline Execution (Zero Data Leakage)
In compliance with Constitution Principle III (*No Data Leakage*), all transformer scaling parameters ($\mu, \sigma$), categorical one-hot dictionaries, and grouping statistics are fitted strictly on `X_train`. The holdout set `X_test` is transformed using the fitted pipeline without parameter re-estimation.

In [8]:
# Fit on training data ONLY
print("Fitting preprocessing pipeline on training data...")
full_pipeline.fit(X_train)

# Transform train and holdout test sets
X_train_proc = full_pipeline.transform(X_train)
X_test_proc = full_pipeline.transform(X_test)

# Extract column names from fitted ColumnTransformer
feature_names = full_pipeline.named_steps['preprocessor'].get_feature_names_out()

# Create clean DataFrames
df_X_train_proc = pd.DataFrame(X_train_proc, columns=feature_names, index=X_train.index)
df_X_test_proc = pd.DataFrame(X_test_proc, columns=feature_names, index=X_test.index)

print("\nPreprocessing Transformation Completed Successfully:")
print(f" - Transformed Training Set: {df_X_train_proc.shape[0]:,} rows × {df_X_train_proc.shape[1]} features")
print(f" - Transformed Test Set:     {df_X_test_proc.shape[0]:,} rows × {df_X_test_proc.shape[1]} features")
print(f" - Total Missing / NaN Values in Train: {df_X_train_proc.isna().sum().sum()}")
print(f" - Total Missing / NaN Values in Test:  {df_X_test_proc.isna().sum().sum()}")

assert df_X_train_proc.shape[1] == 64, f"Expected 64 features, got {df_X_train_proc.shape[1]}"
assert df_X_test_proc.shape[1] == 64, f"Expected 64 features, got {df_X_test_proc.shape[1]}"
assert df_X_train_proc.isna().sum().sum() == 0, "NaN values detected in training set!"
assert df_X_test_proc.isna().sum().sum() == 0, "NaN values detected in test set!"
assert 'duration' not in feature_names, "CRITICAL: 'duration' found in feature names!"
print("Integrity checks passed: 0 missing values, 64 features, and zero leakage.")

Fitting preprocessing pipeline on training data...

Preprocessing Transformation Completed Successfully:
 - Transformed Training Set: 32,950 rows × 64 features
 - Transformed Test Set:     8,238 rows × 64 features
 - Total Missing / NaN Values in Train: 0
 - Total Missing / NaN Values in Test:  0
Integrity checks passed: 0 missing values, 64 features, and zero leakage.


## 6. Post-Scaling Feature Distributions Inspection

In [9]:
print("Sample of Processed Feature Matrix (First 3 rows, selected columns):")
selected_sample_cols = [
    'education', 'previously_contacted', 'pdays_group_0_to_6_days',
    'pdays_group_not_contacted', 'age_group_<30', 'age_group_60+',
    'campaign', 'euribor3m', 'nr.employed'
]
print(df_X_train_proc[selected_sample_cols].head(3))

# Diagnostic Plots
fig, axes = plt.subplots(1, 3, figsize=(16, 5))

# 1. Age distribution standardized
sns.histplot(df_X_train_proc['age'], ax=axes[0], kde=True, color='#2b5c8f', bins=30)
axes[0].set_title("Standardized Age (Mean=0, Std=1)")
axes[0].set_xlabel("Z-Score")

# 2. Campaign distribution robust-scaled
sns.histplot(df_X_train_proc['campaign'], ax=axes[1], kde=False, color='#e07a5f', bins=30)
axes[1].set_title("Robust-Scaled Campaign (Median=0, IQR=1)")
axes[1].set_xlabel("Robust Scaled Value")
axes[1].set_xlim(-1, 10)

# 3. Euribor 3-month rate standardized
sns.histplot(df_X_train_proc['euribor3m'], ax=axes[2], kde=True, color='#81b29a', bins=30)
axes[2].set_title("Standardized Euribor 3M Rate")
axes[2].set_xlabel("Z-Score")

plt.suptitle("Post-Scaling Feature Distributions", fontsize=14, weight='bold')
plt.close(fig)

Sample of Processed Feature Matrix (First 3 rows, selected columns):
       education  previously_contacted  pdays_group_0_to_6_days  pdays_group_not_contacted  age_group_<30  age_group_60+  campaign  euribor3m  nr.employed
25611        4.0                   0.0                      0.0                        1.0            0.0            0.0       1.0   0.288964     0.398990
26010        7.0                   0.0                      0.0                        1.0            0.0            0.0       0.0   0.288964     0.398990
40194        2.0                   0.0                      0.0                        1.0            0.0            1.0      -0.5  -1.583296    -2.420139


## 7. Exporting Processed Dataset Artifacts & Pipeline Serialization
All processed datasets are saved to `data/processed/`, accompanied by the fitted pipeline object and metadata schema.

In [10]:
# Create output directory anchored to PROJECT_ROOT
processed_dir = (PROJECT_ROOT / "data" / "processed") if 'PROJECT_ROOT' in globals() else (Path("data/processed") if Path("data").exists() else Path("../data/processed"))
processed_dir.mkdir(parents=True, exist_ok=True)

# 1. Save processed feature sets
train_features_path = processed_dir / "X_train_processed.csv"
test_features_path = processed_dir / "X_test_processed.csv"
train_target_path = processed_dir / "y_train.csv"
test_target_path = processed_dir / "y_test.csv"

df_X_train_proc.to_csv(train_features_path, index=False)
df_X_test_proc.to_csv(test_features_path, index=False)
y_train.to_csv(train_target_path, index=False)
y_test.to_csv(test_target_path, index=False)

# 2. Serialize full fitted pipeline
pipeline_save_path = processed_dir / "preprocessor_pipeline.joblib"
joblib.dump(full_pipeline, pipeline_save_path)

# 3. Export feature catalog
metadata = {
    "project": "UCI Bank Marketing Campaign Response",
    "phase": "02_Data_Preprocessing_Pipeline",
    "author": "Athukorala M. B. (IT24101937)",
    "git_branch": "002-reconcile-split-strategy",
    "validation_split_strategy": "Option B - Stratified Random Holdout (80/20, seed=42)",
    "decision_log_reference": "docs/decision_log.md#adr-001",
    "total_input_features": X.shape[1],
    "total_output_features": len(feature_names),
    "train_records": len(df_X_train_proc),
    "test_records": len(df_X_test_proc),
    "train_positive_rate": float(y_train.mean()),
    "test_positive_rate": float(y_test.mean()),
    "leakage_feature_dropped": "duration",
    "features_list": list(feature_names)
}

metadata_path = processed_dir / "feature_metadata.json"
with open(metadata_path, 'w') as f:
    json.dump(metadata, f, indent=2)

print("Export completed successfully:")
print(f" - X_train:  {train_features_path} ({train_features_path.stat().st_size / 1024 / 1024:.2f} MB)")
print(f" - X_test:   {test_features_path} ({test_features_path.stat().st_size / 1024 / 1024:.2f} MB)")
print(f" - y_train:  {train_target_path}")
print(f" - y_test:   {test_target_path}")
print(f" - Pipeline: {pipeline_save_path}")
print(f" - Metadata: {metadata_path}")

Export completed successfully:
 - X_train:  /Users/moni/Desktop/ML proj/data/processed/X_train_processed.csv (10.98 MB)
 - X_test:   /Users/moni/Desktop/ML proj/data/processed/X_test_processed.csv (2.75 MB)
 - y_train:  /Users/moni/Desktop/ML proj/data/processed/y_train.csv
 - y_test:   /Users/moni/Desktop/ML proj/data/processed/y_test.csv
 - Pipeline: /Users/moni/Desktop/ML proj/data/processed/preprocessor_pipeline.joblib
 - Metadata: /Users/moni/Desktop/ML proj/data/processed/feature_metadata.json


## 8. Serialization & Inference Verification

In [11]:
# Reload serialized pipeline
loaded_pipeline = joblib.load(pipeline_save_path)

# Take 5 raw samples from X_test (before any transformation)
raw_samples = X_test.head(5)

# Transform using loaded pipeline
transformed_samples = loaded_pipeline.transform(raw_samples)

print(f"Loaded Pipeline Type: {type(loaded_pipeline)}")
print(f"Raw Input Shape:      {raw_samples.shape}")
print(f"Transformed Output:   {transformed_samples.shape}")

# Verify exact numerical match with original transformed test set
original_subset = df_X_test_proc.head(5).to_numpy()
np.testing.assert_allclose(transformed_samples, original_subset, rtol=1e-5, atol=1e-8)
print("Verification Passed: Loaded pipeline transforms raw incoming leads identically!")

Loaded Pipeline Type: <class 'sklearn.pipeline.Pipeline'>
Raw Input Shape:      (5, 19)
Transformed Output:   (5, 64)
Verification Passed: Loaded pipeline transforms raw incoming leads identically!
